### PR TITLE
Support PathBase (#214).

### PR DESCRIPTION
**Pull request by cesarbs**
Using `ServerAddress.Path` as PathBase. Had to change ServerAddress to not allow `Path` to be `/` or end with a `/`, since we don't want that for PathBase.

I'm in doubt about the difference between a request to `/foo` and `/foo/` when PathBase is `/foo`. Should Path be empty in the first case (which OWIN allows) and `/` in the second case, or should it be `/` in both cases? I tried forcing it to be `/` in both cases but that caused some functional tests to fail.

I'd appreciate some input on this.

#214

cc @Tratcher @halter73 @davidfowl

### CODE REVIEW

File: samples/SampleApp/Startup.cs [hunk 1]
```diff
@@ -39,7 +39,7 @@ public void Configure(IApplicationBuilder app, ILoggerFactory loggerFactory, IAp
                 Console.WriteLine("Could not find certificate at '{0}'. HTTPS is not enabled.", testCertPath);
             }
 
-            app.UseKestrelConnectionLogging();
+            //app.UseKestrelConnectionLogging();
```
Comment (cesarbs) [inline samples/SampleApp/Startup.cs]: Revert.

File: samples/SampleApp/project.json [hunk 2]
```diff
@@ -15,8 +15,8 @@
         }
     },
     "commands": {
-        "web": "Microsoft.AspNet.Hosting --server Microsoft.AspNet.Server.Kestrel --server.urls http://localhost:5000;https://localhost:5001",
-        "kestrel": "Microsoft.AspNet.Server.Kestrel --server.urls http://localhost:5000;https://localhost:5001",
+        "web": "Microsoft.AspNet.Hosting --server Microsoft.AspNet.Server.Kestrel --server.urls http://localhost:5000/base;https://localhost:5001",
```
Comment (cesarbs) [inline samples/SampleApp/project.json]: Revert.

**Comment by cesarbs** [comment]
Follow up on my original comment: WebListener sets `Path` to `/` when the request URL ends with a trailing slash, but leaves it empty when it doesn't. That's the current behavior I have for Kestrel in this PR.

File: src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs [hunk 3]
```diff
@@ -744,7 +749,17 @@ private bool TakeStartLine(SocketInput input)
                 RequestUri = requestUrlPath;
                 QueryString = queryString;
                 HttpVersion = httpVersion;
-                Path = RequestUri;
+
+                if (!string.IsNullOrEmpty(_pathBase) && requestUrlPath.StartsWith(_pathBase))
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs]: case insensitive compare, then make sure you set PathBase to the requestUrlPath substring to preserve the user casing.

Also, StartsWith isn't good enough. "/foobar" starts with "/foo".

File: test/Microsoft.AspNet.Server.Kestrel.FunctionalTests/PathBaseTests.cs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+                            Path = context.Request.Path.Value
+                        }));
+                    });
+                });
+
+            using (var app = builder.Build().Start())
+            {
+                using (var client = new HttpClient())
+                {
+                    var response = await client.GetAsync(requestAddress);
```
Comment (Tratcher) [inline test/Microsoft.AspNet.Server.Kestrel.FunctionalTests/PathBaseTests.cs]: HttpClient will always turn http://localhost to http://localhost/, so you can't test that senario here.

File: src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs [hunk 5]
```diff
@@ -744,7 +751,19 @@ private bool TakeStartLine(SocketInput input)
...
                 QueryString = queryString;
                 HttpVersion = httpVersion;
-                Path = RequestUri;
+
+                if (!string.IsNullOrEmpty(_pathBase) &&
+                    (requestUrlPath.Equals(_pathBase, StringComparison.OrdinalIgnoreCase) ||
+                     requestUrlPath.StartsWith(_pathBaseWithSlash, StringComparison.OrdinalIgnoreCase)))
+                {
+                    PathBase = requestUrlPath.Substring(0, _pathBase.Length);
+                    Path = requestUrlPath.Substring(_pathBase.Length);
```
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs]: I'm going to rewrite this to use iterators and also avoid scanning the string multiple times.

File: src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs [hunk 6]
```diff
@@ -744,7 +751,50 @@ private bool TakeStartLine(SocketInput input)
                 RequestUri = requestUrlPath;
                 QueryString = queryString;
                 HttpVersion = httpVersion;
-                Path = RequestUri;
+
+                var pathBaseEnd = pathBegin;
```
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs]: This is not the final revision of it - I just pushed so I can continue working on it later from somewhere else.

**Comment by cesarbs** [comment]
Ping.

File: src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs [hunk 7]
```diff
@@ -71,6 +72,7 @@ public Frame(ConnectionContext context)
             _remoteEndPoint = remoteEndPoint;
             _localEndPoint = localEndPoint;
             _prepareRequest = prepareRequest;
+            _pathBase = context?.ServerAddress?.PathBase;
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs]: Can those really all be null?

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs [hunk 8]
```diff
@@ -203,5 +203,31 @@ public static ArraySegment<byte> GetArraySegment(this MemoryPoolIterator2 start,
             start.CopyTo(array, 0, length, out length);
             return new ArraySegment<byte>(array, 0, length);
         }
+
+        public static bool StartsWithIgnoreCase(this MemoryPoolIterator2 start, string prefix, out MemoryPoolIterator2 end, out bool caseMatches)
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs]: You may want to profile this, we went through extreme lengths to avoid stuff like this on the per request path (in kestrel at least) :smile:

**Comment by Tratcher** [comment]
:+1: on the tests. Good luck with the perf. /cc: @benaadams

File: src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs (hunk 7 above)
Comment (halter73) [inline src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs]: I'm fairly certain (maybe with the exception of some tests) that neither context nor ServerAddress can be null.

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs [hunk 9]
```diff
@@ -203,5 +203,31 @@ public static ArraySegment<byte> GetArraySegment(this MemoryPoolIterator2 start,
...
+            caseMatches = true;
+
+            var i = 0;
+            for (i = 0; i < prefix.Length; i++)
+            {
+                var c = end.Take();
+
+                if (c != prefix[i])
+                {
+                    if (char.ToLowerInvariant((char)c) == char.ToLowerInvariant(prefix[i]))
```
Comment (halter73) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs]: I wonder if you could improve perf for the ASCII case by writing your own `ToLowerInvariantAscii` or something like that.

Comment (benaadams) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs]: `Vector.BitwiseOr(new Vector<byte>(0x20), v)` ...

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs (hunk 8 above)
Comment (halter73) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs]: :+1: for profiling. @davidfowl Do you have any specific ideas for improving the perf of these PR?

**Comment by halter73** [comment]
:+1: on the tests

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs (hunk 9 above)
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs]: @halter73 char.ToLowerInvariant already optimizes the call when it's ASCII. It ends up here:

http://referencesource.microsoft.com/#mscorlib/system/globalization/textinfo.cs,145c4c32b3b4bace

File: src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs [hunk 10]
```diff
@@ -744,7 +748,31 @@ private bool TakeStartLine(SocketInput input)
...
                 QueryString = queryString;
                 HttpVersion = httpVersion;
-                Path = RequestUri;
+
+                MemoryPoolIterator2 pathBaseEnd;
+                bool caseMatches;
+
+                if (!string.IsNullOrEmpty(_pathBase) &&
+                    requestUrlPath.Length >= _pathBase.Length &&
+                    pathBegin.StartsWithIgnoreCase(_pathBase, out pathBaseEnd, out caseMatches) &&
```
Comment (benaadams) [inline src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs]: if _pathBase is a fixed string/char array; isn't it excessive to create a MemoryPoolIterator2 out of it for the purpose of comparison? Could also create a _pathBaseLower at Frame initialisation time; rather than doing it on each request per char again and again?

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs [hunk 11]
```diff
@@ -203,5 +203,31 @@ public static ArraySegment<byte> GetArraySegment(this MemoryPoolIterator2 start,
...
             return new ArraySegment<byte>(array, 0, length);
         }
+
+        public static bool StartsWithIgnoreCase(this MemoryPoolIterator2 start, string prefix, out MemoryPoolIterator2 end, out bool caseMatches)
+        {
+            end = start;
+            caseMatches = true;
+
+            var i = 0;
+            for (i = 0; i < prefix.Length; i++)
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs]: Does the iterator have a way of scanning length chars?

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs (hunk 9 above)
Comment (halter73) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs]: @cesarbs that's a little more efficient than I thought, but that checks if each character is in the ascii range, which might be unnecessary. Could we try @benaadams suggestion and vectorize the operation to see if it performs better?

File: src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs (hunk 10 above)
Comment (halter73) [inline src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs]: We would like to know if the case matches too. If so, we don't need to allocate a new string and can just do `PathBase = _pathBase`. Otherwise, we have to create a new correctly cased string.

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs (hunk 11 above)
Comment (halter73) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs]: I don't think that exists yet. How do you expect that to work? You pass a length and get an `IEnumerable<char>`? Something callback based?

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs (hunk 9 above)
Comment (Tratcher) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs]: Keep in mind that the path has already been un-escaped, so `c` may be utf-8 bytes. @cesarbs add a test for base paths that contain unicode characters: `/b𤭢se`

Comment (halter73) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs]: Correct, but only if `needDecode` is true. We can have a special ASCII path if we want to.

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs (hunk 11 above)
Comment (davidfowl) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs]: `ArraySegment<byte>`

Comment (halter73) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs]: Would a new array be allocated if the chars traverse blocks?

Comment (davidfowl) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs]: I guess that would kinda suck huh :smile:

**Comment by cesarbs** [comment]
I've stepped away from using iterators to do the path base comparison. The code started looking quite messy when I included UTF-8 characters in the game.

Since I already have a decoded string for the whole request URL, I'm simply using that to compare against _pathBase. I did a quick check to make sure that the new code doesn't significantly impact our requests per seconds and it's not the case.

I see little advantage in using iterators anyways, since what we're trying to optimize for is reduced string allocations. With iterators or not we still have to allocate a new string if the path base in the request doesn't match case with the registered path base. And we will always have to allocate a new string for the path.

**Comment by halter73** [comment]
Squash this.

**Comment by cesarbs** [comment]
@halter73 Done :+1:

**Comment by halter73** [comment]
@davidfowl @cesarbs explained to me that doing an iterator based comparison against pathbase made no measurable improvement in RPS (using wrk) over the PR as it currently stands.

Given your aversion to complicating the codebase for no rps/memory gains, I assume your fine with taking this PR as is? Technically, we might be able to reduce allocations if the full path is exactly equal to the PathBase, but I'm not sure if that is worth special casing to remove one string allocation per impacted request.

We could always look at improving the performance of this down the road.

**Comment by halter73** [comment]
:shipit:

**Comment by davidfowl** [comment]
:shipit: 

/cc @DamianEdwards measure when this is merged
